src/nvim/buffer.c

```diff
@@ -309,7 +309,7 @@ close_buffer (
   // buffers.
   // The caller must take care of NOT deleting/freeing when 'bufhidden' is
   // "hide" (otherwise we could never free or delete a buffer).
-  if (!buf->terminal) {
+  if (!buf->terminal && !buf->b_messages) {
```

Comment:
I'm going to look at removing the special treatment after I sort out how the lines are added to the buffer.  As it is now, if a user happens to add a line after the buffer reaches 1000 lines, the buffer will grow indefinitely.